[MERGE-AFTER-PR-5366] Multi-container workspace Support

davidfestal opened the pull request:
### What does this PR do?

Fixes the support of multi-container (multi-machine) Che workspaces inside Openshift.

In the current design of the OpenShift connector, each Che machine create is transformed into an OpenShift Pod with a single container. (and not a Pod with many containers).

However, the multi-machine use-case wasn't fully tested and NPEs used to prevent the setup of non-development additional machines. This is now fixed.

Additionally, since machines live in distinct pods, the docker compose internal network feature cannot be used to allow machines communicating with their compose service names.
To fix this, the OpenShift service name, for all non-development Che machines, will be created with the name of the related compose service. This way we can leverage the OpenShift internal Dns server that allows all pods in an OS namespace to communicate through their OS service name.

### What issues does this PR fix or reference?

https://issues.jboss.org/browse/CHE-175
https://issues.jboss.org/browse/CHE-258
https://issues.jboss.org/browse/CHE-259

davidfestal commented:
@l0rd, @gorkem, could you review this PR please ?

l0rd commented on plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/kubernetes/KubernetesLabelConverter.java:
```diff
@@ -64,31 +64,33 @@ public static String getCheServerLabelPrefix() {
-                    LOG.error(
-                            "Could not convert label {} into Kubernetes annotation: labels must be alphanumeric with ':' and '/'",
-                            label.toString());
+        if (labels != null) {
```
Avoiding to nest the for loop inside the `if` would make it more readable:
```  java
if (labels == null) {
  returns names;
}

for (Map.Entry<String, String> label : labels.entrySet()) {
 ...
}
```

l0rd commented on plugins/plugin-docker/che-plugin-openshift-client/src/main/java/org/eclipse/che/plugin/openshift/client/kubernetes/KubernetesLabelConverter.java:
```diff
@@ -103,25 +105,27 @@ public static String getCheServerLabelPrefix() {
-            }
-            key = keyMatcher.group(1);
-            value = valueMatcher.group(1);
+        if (names != null) {
```
Same comment here than above.

l0rd commented:
cc @sunix

davidfestal commented on ide/che-core-ide-stacks/src/main/resources/stacks.json:
```diff
@@ -1,5 +1,74 @@
 [
   {
+    "id": "java-centos-mysql",
```
I don't know. @l0rd can tell.

l0rd replied:
@davidfestal @sunix it makes sense to me to merge this stack to eclipse/che master

davidfestal commented:
@l0rd I committed e66d31df1b29033b4ccff9ca9d7ac9c123a28a20 as you suggested. Do you think you could you merge ?

sunix commented on ide/che-core-ide-stacks/src/main/resources/stacks.json:
```diff
@@ -1,5 +1,74 @@
+          "machines": {
+            "dev-machine": {
+              "agents": [
+                "org.eclipse.che.terminal", "org.eclipse.che.ws-agent", "com.redhat.bayesian.lsp"
```
if this is supposed to go to master, it shouldn't include com.redhat.bayesian.lsp agent

l0rd replied:
That's a good point @sunix. I think @davidfestal did it for consistency since other stacks use the same agent here. But I understand that's an issue for you because you are in the middle of a rebasing. I think we have 2 choices:
- @davidfestal remove `com.redhat.bayesian.lsp` and we merge it right now
- @davidfestal change the PR to be merge on `master` branch instead of `openshift-connector`

@sunix are these 2 options ok for you?
@davidfestal which of these 2 options is better for you?

l0rd replied:
When I say change the PR to be merged on `master` I actually mean to split this PR in two:
- one PR for `master` with the modifications of `stacks.json`
- one PR for `openshift-connector` with modifications on the `OpenShiftConnector`

davidfestal commented:
@l0rd 

> remove `com.redhat.bayesian.lsp` and we merge it right now

Let me choose this option. Then we'll be able to merge on `openshift-connector` branch right now, right ?

davidfestal commented:
@l0rd @sunix I changed the stacks file to remove the default activation of Bayesian. Is it OK to merge it ?

sunix commented:
The second problem is that in the branch `openshift-connector-rebased` and `openshift-connector-wip` we did reverted all changes on stack.json, so if you could cherry-pick and fix conflicts + send PR (to openshift-connector-rebased) it would work for me

davidfestal commented:
@sunix. Not sure I fully understand what you're proposing. But let me remove the `stacks.json` from this PR and simply add the `stacks.json` change in a PR against master, since @l0rd said the newly-created `java-mysql-centos` stack was expected to finally be merged to `eclipse/che` `master`. This shouldn't bring any conflict like that, right ? Of course we'll have to reintegrate this new stack into the `rh-che` `stacks.json` file.

@l0rd wdyt ?

davidfestal commented:
@sunix I removed the `stacks.json` from this PR and  created a new PR for it (#5227).
@l0rd I added the changes that also fix https://issues.jboss.org/browse/CHE-258 and https://issues.jboss.org/browse/CHE-259

Is it possible to merge this ?

l0rd commented:
@davidfestal can you please rebase this on top of branche openshift-connector-rebased

davidfestal commented:
@l0rd @sunix I rebased to the `openshift-connector-rebased` branch

However this should be rebuilt and tested again when merged to the `openshift-connector-branch` or to `master`

davidfestal commented:
@sunix could you review this PR ? it should be able to be merged without any problem / conflicts now.

benoitf commented:
maybe in the description, it could say that each Eclipse Che machine is transformed into an OpenShift Pod with a single container. (and not a Pod with many containers)

davidfestal commented:
> maybe in the description, it could say that each Eclipse Che machine is transformed into an OpenShift Pod with a single container. (and not a Pod with many containers)

Sure let me add a more detailed explanation
